components/mongodb/mongodb.app.js

```diff
@@ -0,0 +1,82 @@
+const MongoClient = require("mongodb").MongoClient;
+
+module.exports = {
+  type: "app",
+  app: "mongodb",
+  propDefinitions: {
+    database: {
+      type: "string",
+      label: "Database",
+      async options() {
+        const client = await this.getClient();
+        const databases = await this.listDatabases(client);
+        await client.close();
+        await new Promise((resolve) => setTimeout(resolve, 3000));
+        return databases.map((db) => {
+          return db.name;
+        });
+      },
+    },
+    collection: {
+      type: "string",
+      label: "Collection",
+      async options(opts) {
+        const client = await this.getClient();
+        const db = this.getDatabase(client, opts.database);
+        const collections = await this.listCollections(db);
+        await client.close();
+        await new Promise((resolve) => setTimeout(resolve, 3000));
+        return collections.map((collection) => {
+          return collection.name;
+        });
+      },
+    },
+    document: {
+      type: "string",
+      label: "Document",
+      async options(opts) {
+        const client = await this.getClient();
+        const db = this.getDatabase(client, opts.database);
+        const collection = this.getCollection(db, opts.collection);
+        const documents = await this.listDocuments(collection);
+        await client.close();
+        await new Promise((resolve) => setTimeout(resolve, 3000));
+        return documents.map((doc) => {
+          return doc._id;
+        });
+      },
+    },
+  },
+  methods: {
+    _getConnectionUrl() {
+      const {
+        username,
+        password,
+        hostname,
+      } = this.$auth;
+      return `mongodb+srv://${username}:${password}@${hostname}`;
+    },
+    async getClient() {
+      const url = this._getConnectionUrl();
+      return await MongoClient.connect(url);
+    },
```

Comment:
This connection URI won't work in all cases (e.g. when a user has a single MongoDB instance with just an IP address):
```
> Error: querySrv ENOTFOUND _mongodb._tcp.34.82.65.182
    at QueryReqWrap.onresolve [as oncomplete] (dns.js:203:19)
    at QueryReqWrap.callbackTrampoline (internal/async_hooks.js:134:14) {
  errno: undefined,
  code: 'ENOTFOUND',
  syscall: 'querySrv',
  hostname: '_mongodb._tcp.34.82.65.182'
}
```

I suggest the following so that we cover this case as well as the case with replica sets:
```suggestion
    _getConnectionUrl(replSet = true) {
      const {
        username,
        password,
        hostname,
      } = this.$auth;
      const protocol = replSet
        ? "mongodb+srv"
        : "mongodb";
      return `${protocol}://${username}:${password}@${hostname}`;
    },
    async getClient() {
      const url = this._getConnectionUrl();
      try {
        return await MongoClient.connect(url);
      } catch (err) {
        if (err.code === "ENOTFOUND") {
          // Last attempt to connect to the server
          const url = this._getConnectionUrl(false);
          return MongoClient.connect(url);
        }

        throw err;
      }
    },
```

Reply:
Updated